harness/determined/pytorch/_reducer.py

```diff
@@ -184,7 +185,21 @@ def per_slot_reduce(self) -> Any:
         return self.values
 
     def cross_slot_reduce(self, per_slot_metrics: List) -> Any:
-        flat_metrics = [item for sublist in per_slot_metrics for item in sublist]
+
+        # restore the initial (pre-sharding) order of metrics
```

Comment:
This is a perfect place to put the explanation of assumptions and limitations that you have in the PR comment, I think.

Also, I have never used it, but I checked if `itertools` has a tool for this style of interleaved iteration and it turns that out [it does](https://docs.python.org/3/library/itertools.html#itertools.zip_longest)